test: add catch all mock and temporary allowlist

seaona opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**
This PR does the following:
- adds a catch all mock, which will prevent any test to send real requests to an API, and instead returns a 200
- displays logs anytime the request is going to a live server, or redirected to the catch all mock--> for easy debugging
- adds a url and domain allowlist, which allows the requests to those entries to be performed in the real servers. This list is temporary and the goal is to reduce this to 0. However, adding this list was necessary in order to make the specs that rely on those current live responses to not fail and be able to merge this PR.

Merging this PR will make that any new request, is catched by the catch all mock, so it already prevents new requests to happen.
The work on mocking all the whitelisted requests should be a joint effort between teams.


[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/26647?quickstart=1)

## **Related issues**

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/2762

## **Manual testing steps**

1. Check ci should work

## **Screenshots/Recordings**
When a request is sent to a live server, we'll see the following information in the test logs:

![Screenshot from 2025-03-20 08-17-30](https://github.com/user-attachments/assets/a66d56a7-c4fe-4564-8f9f-3b60f80d2526)




I've enabled all test jobs just to be sure it doesn't break anything

https://app.circleci.com/pipelines/github/MetaMask/metamask-extension/132122/workflows/4ba5987a-28c3-41c3-8d5a-312dc88aacce

![Screenshot from 2025-03-19 16-57-18](https://github.com/user-attachments/assets/fc85f977-fbf6-4997-b9c6-79bd563fe919)

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

seaona committed:
catch all mock

gauthierpetetin commented:
Notes from refinement session:

- We can't merge this PR because some tests are failing (more than a hundred), due to not being mocked
- We don't know which ones are missing mocks

Suggested next steps:

- [ ] Investigate which mocks are missing
- [ ] For each missing mock, decide whether we want to add it or not. Ideally we decide to mock them all. However, some of these are owned by other teams and we can't always decide on their behalf.
  - [ ] If we decide to mock it, create an issue for it
  - [ ] If we decide not to mock it, add an exception in this PR

seaona committed:
add whitelisted mocks

seaona committed:
add comments

seaona committed:
add debugging for missing whitelisted urls

seaona committed:
add 2 more whitelisted endpoints

seaona committed:
more urls to the whitelist

seaona committed:
more urls

seaona committed:
added more hosts and urls

seaona committed:
update list

Montoya committed:
Docs: Update Developer Docs link in README (#29733)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

This link was pointing to a dead site, this fixes it to point to the
correct URL.

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/29733?quickstart=1)

## **Related issues**

N/A

## **Manual testing steps**

1. Go to
https://github.com/MetaMask/metamask-extension/blob/develop/docs/README.md
2. Click the link to Developer Docs 

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask
Extension Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

seaona committed:
Merge branch 'develop' into test-catch-all-mock

seaona committed:
Merge branch 'main' into test-catch-all-mock

seaona committed:
thielist update

seaona committed:
more whitelist updates

seaona committed:
more updates

seaona committed:
mocks

seaona committed:
solana

seaona committed:
more whitelisted

seaona committed:
run specs in all jobs to test

seaona committed:
more whitelisted

seaona committed:
bundle

seaona committed:
fix

seaona committed:
add filters back

seaona commented on test/e2e/mock-e2e-whitelisted.js:
```diff
@@ -0,0 +1,192 @@
+  'https://www.4byte.directory/api/v1/signatures/?hex_signature=0xd0e30db0',
+];
+
+const WHITE_LISTED_HOSTS = [
```
I tried to be granular, adding each specific URL in the above list, but for the cases where there were a huge amount of requests to different URLS with the same domain, or cases where the URL changed for some reason (including random ids etc) then it has been added as a host in the below list.

Both lists should be reduced to 0 eventually

chloeYue reviewed (dismissed):
LGTM !

seaona committed:
add logs and redirect

seaona committed:
url

seaona committed:
add to codeowners

seaona commented on .github/CODEOWNERS:
```diff
@@ -137,4 +137,5 @@ ui/components/ui/deprecated-networks                  @MetaMask/metamask-assets
-yarnrc.yml                                            @MetaMask/extension-platform
\ No newline at end of file
+yarnrc.yml                                            @MetaMask/extension-platform
+mock-e2e-whitelisted.js                               @MetaMask/extension-platform
```
:information_source:  I've added this into the platform team codeowners, for our team to ensure that no new urls/hosts are added in the whitelisted file (only removed)

seaona replied:
ℹ️ I've added this into the platform team codeowners, for our team to ensure that no new urls/hosts are added in the whitelisted file (only removed)

davidmurdoch commented on .github/CODEOWNERS:
```diff
@@ -137,4 +137,5 @@ ui/components/ui/deprecated-networks                  @MetaMask/metamask-assets
-yarnrc.yml                                            @MetaMask/extension-platform
\ No newline at end of file
+yarnrc.yml                                            @MetaMask/extension-platform
+mock-e2e-whitelisted.js                               @MetaMask/extension-platform
```
I think this should be `test/e2e/mock-e2e-whitelisted.js`

seaona replied:
awwww  :crying_cat_face: thank you!!!

Gudahtt commented on test/e2e/mock-e2e.js:
```diff
@@ -117,14 +122,26 @@ async function setupMocking(
+      console.log('Request redirected to a catch all mock ============', url);
+      return {
+        // If the URL is not whitelisted nor blacklisted, we redirect the request to the localhost fixture server
+        // to always return a 200 response.
```
😄 Well that's one way to deal with them.

Perhaps instead of redirecting to this server, we could return the response here directly? The fixture data doesn't seem especially helpful to return here.

We can return an object with a `response` property to respond to the request directly here, without it being passed through further (see https://httptoolkit.github.io/mockttp/interfaces/requestHandlers.CallbackRequestResult.html#response)

Gudahtt replied:
I'm also curious, why a successful response rather than a failure? Just to minimize disruption?

Maybe this is the best idea as a first step, but eventually it would seem helpful to track these in some way so we can more easily catch them. And even though this is logged, it may be easy to miss (people usually don't read these logs until something fails).

Gudahtt replied:
I am unsure on the best approach as well, but going with what you've suggested here (a successful 200 response) seems like an OK decision for now. It's easy to change it later if we change our minds.

seaona replied:
it make sense to me! Maybe we can proceed with this 200 as the first step, then, we'll work on reducing the allowlist to 0 and as a later step, review what to do with the catch-all bucket again

Gudahtt commented on test/e2e/mock-e2e.js:
```diff
@@ -20,6 +20,11 @@ const {
+const {
+  WHITE_LISTED_HOSTS,
+  WHITE_LISTED_URLS,
+} = require('./mock-e2e-whitelisted');
```
Nit: We should prefer using more inclusive language, "allowlist" over "whitelist". I see "backlist" is already used though, so maybe this was meant to follow that pre-existing convention. But we already use blocklist/allowlist in most places, e.g. the phishing config.

seaona replied:
sure make sense! Changing it now :pray:

seaona committed:
address review comments

seaona committed:
rename

seaona committed:
address comments

chloeYue reviewed (approved):
I went through all the discussions in this PR, great discussion ! 
LGTM to merge !

davibroc reviewed (approved):
LGTM!

Gudahtt reviewed (approved):
LGTM!
